src/java.desktop/macosx/classes/sun/awt/CGraphicsEnvironment.java

```diff
@@ -163,44 +156,74 @@ protected void finalize() throws Throwable {
     /**
      * (Re)create all CGraphicsDevices, reuses a devices if it is possible.
      */
-    private void initDevices() {
-        synchronized (this) {
-            final Map<Integer, CGraphicsDevice> old = new HashMap<>(devices);
-            devices.clear();
-
-            mainDisplayID = getMainDisplayID();
-
-            // initialization of the graphics device may change
-            // list of displays on hybrid systems via an activation
-            // of discrete video.
-            // So, we initialize the main display first, and then
-            // retrieve actual list of displays.
-            if (!old.containsKey(mainDisplayID)) {
-                old.put(mainDisplayID, new CGraphicsDevice(mainDisplayID));
+    private synchronized void initDevices() {
+        Map<Integer, CGraphicsDevice> old = new HashMap<>(devices);
+        devices.clear();
+        mainDisplayID = getMainDisplayID();
+
+        // initialization of the graphics device may change list of displays on
+        // hybrid systems via an activation of discrete video.
+        // So, we initialize the main display first, then retrieve actual list
+        // of displays, and then recheck the main display again.
+        if (!old.containsKey(mainDisplayID)) {
+            old.put(mainDisplayID, new CGraphicsDevice(mainDisplayID));
+        }
+
+        int[] displayIDs = getDisplayIDs();
+        if (displayIDs.length == 0) {
+            // we could throw AWTError in this case.
+            displayIDs = new int[]{mainDisplayID};
+        }
+        for (int id : displayIDs) {
+            devices.put(id, old.containsKey(id) ? old.remove(id)
+                                                : new CGraphicsDevice(id));
+        }
+        // fetch the main display again, the old value might be outdated
+        mainDisplayID = getMainDisplayID();
```

Comment:
What action since line 162 can trigger changing the getMainDisplayID() outcome?

Reply:
If the main screen is not in the list of screens means that the system monitors were changed in some way after this method was called and we will get one more notification later.